### PR TITLE
rewrite checked_{div,rem} to no contain any reference to panics

### PR DESCRIPTION
**Pull request by japaric**
even without optimizations

r? @alexcrichton

### CODE REVIEW

**Comment by japaric** [comment]
While implementing this. I noticed that `overflowing_neg` and `checked_neg` also have this problem, they directly use `-self`, but there's no `unchecked_neg` intrinsic to replace that with.

**Comment by alexcrichton** [comment]
Hm I don't think `-foo` suffers the same problems, the compiler doesn't emit a call to panic if that overflows because it can't be UB. (in LLVM I think it's UB to divide by zero, so the panic is necessary for safety)

**Comment by japaric** [comment]
@alexcrichton Oh, that makes sense.

File: src/libcore/num/mod.rs [hunk 1]
```diff
@@ -516,11 +516,10 @@ macro_rules! int_impl {
         #[stable(feature = "rust1", since = "1.0.0")]
         #[inline]
         pub fn checked_div(self, other: Self) -> Option<Self> {
-            if other == 0 {
+            if other == 0 || (self == Self::min_value() && other == -1) {
                 None
             } else {
-                let (a, b) = self.overflowing_div(other);
-                if b {None} else {Some(a)}
+                Some(unsafe { intrinsics::unchecked_div(self, other) })
```
Comment (nagisa) [inline src/libcore/num/mod.rs]: Why intrinsic, not `wrapping_div`?

**Pull request merged**